custom_components/ttlock/services.py

```diff
@@ -138,18 +207,59 @@ async def handle_create_passcode(self, call: ServiceCall):
             endDate=end_time,
         )
 
-        for coordinator in self._get_coordinators(call):
+        for _entity_id, coordinator in self._get_coordinators(call):
             await coordinator.api.add_passcode(coordinator.lock_id, config)
 
     async def handle_cleanup_passcodes(self, call: ServiceCall) -> ServiceResponse:
         """Clean up expired passcodes for the given entities."""
-        removed = []
+        removed = {}
 
-        for coordinator in self._get_coordinators(call):
+        for entity_id, coordinator in self._get_coordinators(call):
+            removed_for_lock = []
             codes = await coordinator.api.list_passcodes(coordinator.lock_id)
             for code in codes:
                 if code.expired:
-                    await coordinator.api.delete_passcode(coordinator.lock_id, code.id)
-                    removed.append(code.name)
+                    if await coordinator.api.delete_passcode(
+                        coordinator.lock_id, code.id
+                    ):
+                        removed_for_lock.append(code.name)
+            if removed_for_lock:
+                removed[entity_id] = removed_for_lock
 
         return {"removed": removed}
+
+    async def handle_list_records(self, call: ServiceCall) -> ServiceResponse:
+        """List records for the selected locks."""
+        records = {}
+        params = {}
+
+        # Convert datetime parameters to millisecond timestamps if provided
+        if start_date := call.data.get("start_date"):
+            params["start_date"] = int(as_utc(start_date).timestamp() * 1000)
+        if end_date := call.data.get("end_date"):
+            params["end_date"] = int(as_utc(end_date).timestamp() * 1000)
+
+        # Get pagination values from call data with defaults
+        params["page_no"] = call.data.get("page_no", 1)
+        params["page_size"] = min(call.data.get("page_size", 50), 200)
+
+        for entity_id, coordinator in self._get_coordinators(call):
+            lock_records = await coordinator.api.get_lock_records(
+                coordinator.lock_id, **params
+            )
+            records[entity_id] = [
+                {
+                    "id": record.id,
+                    "lock_id": record.lock_id,
+                    "record_type": record.record_type.name,
+                    "record_type_from_lock": record.record_type_from_lock,
+                    "success": record.success,
+                    "username": record.username,
+                    "keyboard_pwd": record.keyboard_pwd,
+                    "lock_date": record.lock_date,
+                    "server_date": record.server_date,
+                }
+                for record in lock_records
+            ]
```

Comment:
Out of curiosity, does the json serialiser do what you want if you just present it as follows? 
```suggestion
            records[entity_id] = lock_records
```

Reply:
Sadly not. I did try to do it like you mention but I get "Failed to perform the action ttlock.list_records. Invalid JSON in response". 

2025-01-10 08:49:25.906 ERROR (MainThread) [homeassistant.components.websocket_api.messages] Unable to serialize to JSON. Bad data found at $.result.response.records.lock.lock_box[0]=id=1207157166 lock_id=15450395 record_type_from_lock=7 record_type=<RecordType.PASSWORD_UNLOCK: 4> success=False username='0***' keyboard_pwd='0608' lock_date=datetime.datetime(2025, 1, 6, 15, 1, 58, tzinfo=zoneinfo.ZoneInfo(key='Europe/Bucharest')) server_date=datetime.datetime(2025, 1, 6, 15, 0, 42, tzinfo=zoneinfo.ZoneInfo(key='Europe/Bucharest'))
(<class 'custom_components.ttlock.models.LockRecord'>, $.result.response.records.lock.lock_box[1]=id=1207156760 lock_id=15450395 record_type_from_lock=7 record_type=<RecordType.PASSWORD_UNLOCK: 4> success=False username='0***' keyboard_pwd='0608' lock_date=datetime.datetime(2025, 1, 6, 15, 1, 45, tzinfo=zoneinfo.ZoneInfo(key='Europe/Bucharest')) server_date=datetime.datetime(2025, 1, 6, 15, 0, 28, tzinfo=zoneinfo.ZoneInfo(key='Europe/Bucharest'))

This is basically because we need to convert the record_type to convert it to what we need. We also need to take care of the lock_date and server_date. 

Based on the response I posted here please let me know if you see a way to do it differently and I will for sure try ;)

Reply:
Ok thanks - probably fine to leave it as is then.

If you wanted to get carried away, pydantic supports custom serialisation logic:
https://docs.pydantic.dev/latest/concepts/serialization/#dictmodel-and-iteration

Would probably look something like this: 
```
    @field_serializer('record_type')
    def serialize_record_type(self, record_type: RecordType, _info):
        return record_type.name
```

and then if the home assistant serialiser still can't figure it out, you probably need to do 
```
records[entity_id] = [r.model_dump() for r in lock_records]
```

but I think from memory that part happens automatically.

either way, this is more a nicety than a requirement, I'll happily accept the PR without this change.

Reply:
I looked a bit at this but tbh I am out of my depth here and would not want to do it blindly. I am open to any change on the branch if you want and I can test it on my instance if this is something you would want to try. Sorry for not being more knowledgeable yet on this

Reply:
All good, I can poke around at this later in a separate PR.